File: sieve.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"math"
)

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Please enter one or more arguments, an integer limit",
			"to the size of the primes generated.")
		return
	}

	smart := false
	smart2 := false
	smart3 := false
	quiet := false
	n := 0

	for i := 1; i < len(os.Args); i++ {
		if os.Args[i] == "-s" {
			smart = true
			continue
		}
		if os.Args[i] == "-s2" {
			smart2 = true
			continue
		}
		if os.Args[i] == "-s3" {
			smart3 = true
			continue
		}
		if os.Args[i] == "-q" {
			quiet = true
			continue
		}
		nTemp, err := strconv.Atoi(os.Args[i])
		if err != nil {
			fmt.Println(err)
			os.Exit(2)
		}
		n = nTemp
	}

	if n < 2 {
		fmt.Println("The integer limit must be greater than or equal to two.")
		os.Exit(2)
	}

	if smart3 {
		smart_primes3(n, quiet)
	} else if smart2 {
		smart_primes2(n, quiet)
	} else if smart {
		smart_primes(n, quiet)
	} else {
		dumb_primes(n, quiet)
	}
}

func dumb_primes(n int, quiet bool) {
	var primes []int

	for i := 2; i <= n; i++ {
		primes = append(primes, i)
	}

	for i := 0; i < len(primes); i++ {
		if primes[i] != 0 {
			sieve(primes, primes[i])
		}
	}

	for i := 0; i < len(primes); i++ {
		if primes[i] != 0 {
			if !quiet {
				fmt.Println(primes[i])
			}
		}
	}
}

func smart_primes(n int, quiet bool) {
	primes := make([]bool, n)

	for i := 0; i < n; i++ {
		primes[i] = true
	}

	for factor := 2; factor < n; factor++ {
		if primes[factor] {
			var multiple = 2
			for factor*multiple < n {
				primes[factor*multiple] = false
				multiple += 1
			}
		}
	}

	for i := 0; i < len(primes); i++ {
		if primes[i] {
			if !quiet {
				fmt.Println(i)
			}
		}
	}
}

func smart_primes2(n int, quiet bool) {
	primes := make([]bool, n)

	for i := 0; i < n; i++ {
		primes[i] = true
	}

	slimit := int(math.Sqrt(float64(n)))
	for factor := 2; factor < slimit; factor++ {
		if primes[factor] {
			var multiple = 2
			for factor*multiple < n {
				primes[factor*multiple] = false
				multiple += 1
			}
		}
	}

	for i := 0; i < len(primes); i++ {
		if primes[i] {
			if !quiet {
				fmt.Println(i)
			}
		}
	}
}

func smart_primes3(n int, quiet bool) {
	primes := make([]bool, n)

	for i := 0; i < n; i++ {
		primes[i] = true
	}

	slimit := int(math.Sqrt(float64(n)))
	for factor := 2; factor < slimit; factor++ {
		if primes[factor] {
			var multiple = factor
			for factor*multiple < n {
				primes[factor*multiple] = false
				multiple += 1
			}
		}
	}

	for i := 0; i < len(primes); i++ {
		if primes[i] {
			if !quiet {
				fmt.Println(i)
			}
		}
	}
}

func sieve(primes []int, factor int) {
	for index, value := range primes {
		if value != 0 && value != factor {
			if value%factor == 0 {
				primes[index] = 0
			}
		}
	}
}

File: src/sieve.rs
fn main() {
    let args: Vec<String> = std::env::args().collect();

    if args.len() <= 2 {
        println!(
            "Please enter one or more arguments, an integer limit \
            to the size of the primes generated."
        );
        return;
    }

    let smart = args.iter().any(|arg| arg == "-s");
    let smart2 = args.iter().any(|arg| arg == "-s2");
    let quiet = args.iter().any(|arg| arg == "-q");

    let n: usize = args
        .iter()
        .fold(None, |n, arg| {
            if n.is_none() {
                arg.parse::<usize>().ok()
            } else {
                n
            }
        })
        .expect("No valid unsigned integer argument found");

    if n < 2 {
        println!("The integer limit must be greater than or equal to two.")
    }

    let primes = if smart2 {
        smart_primes2(n)
    } else if smart {
        smart_primes(n)
    } else {
        dumb_primes(n)
    };

    if !quiet {
        for prime in primes {
            println!("{}", prime);
        }
    }
}

fn dumb_primes(n: usize) -> Vec<usize> {
    fn sieve(primes: &mut Vec<usize>, factor: usize) {
        for i in 0..primes.len() {
            let value = primes[i];
            if value != 0 && value != factor {
                if value % factor == 0 {
                    primes[i] = 0;
                }
            }
        }
    }

    let mut primes: Vec<usize> = Vec::new();

    for i in 2..=n {
        primes.push(i);
    }

    for i in 0..primes.len() {
        let factor = primes[i];
        if factor != 0 {
            sieve(&mut primes, factor);
        }
    }

    primes
        .iter()
        .filter_map(|b| if *b != 0 { Some(*b) } else { None })
        .collect()
}

fn smart_primes(n: usize) -> Vec<usize> {
    let mut primes = vec![true; n];
    primes[0] = false;
    primes[1] = false;

    for factor in 2..n {
        if primes[factor] {
            let mut multiple = 2;
            while factor * multiple < n {
                primes[factor * multiple] = false;
                multiple += 1;
            }
        }
    }

    primes
        .iter()
        .enumerate()
        .filter_map(|(i, b)| if *b { Some(i) } else { None })
        .collect()
}

fn smart_primes2(n: usize) -> Vec<usize> {
    let mut primes = vec![true; n];
    primes[0] = false;
    primes[1] = false;

    for factor in 2..(n as f64).sqrt() as usize {
        if primes[factor] {
            let mut multiple = 2;
            while factor * multiple < n {
                primes[factor * multiple] = false;
                multiple += 1;
            }
        }
    }

    primes
        .iter()
        .enumerate()
        .filter_map(|(i, b)| if *b { Some(i) } else { None })
        .collect()
}


#[test]
fn test_sieves() {
    assert_eq!(dumb_primes(10000), smart_primes(10000));
}

#[test]
fn test_sieves2() {
    assert_eq!(dumb_primes(10000), smart_primes2(10000));
}
